gossip/src/received_cache.rs

```diff
@@ -0,0 +1,185 @@
+
+    pub(crate) fn record(&mut self, origin: Pubkey, node: Pubkey, num_dups: usize) {
+        match self.0.get_mut(&origin) {
+            Some(entry) => entry.record(node, num_dups),
+            None => {
+                let mut entry = ReceivedCacheEntry::default();
+                entry.record(node, num_dups);
+                self.0.put(origin, entry);
+            }
+        }
+    }
+
+    pub(crate) fn prune(
+        &mut self,
+        pubkey: Pubkey, // This node.
+        origin: Pubkey, // CRDS value owner.
+        stake_threshold: f64,
+        min_ingress_nodes: usize,
+        stakes: &HashMap<Pubkey, u64>,
+    ) -> impl Iterator<Item = Pubkey> {
+        match self.0.peek_mut(&origin) {
+            None => None,
+            Some(entry) if entry.num_upserts < Self::MIN_NUM_UPSERTS => None,
+            Some(entry) => Some(
+                std::mem::take(entry)
+                    .prune(&pubkey, &origin, stake_threshold, min_ingress_nodes, stakes)
+                    .filter(move |node| node != &origin),
+            ),
+        }
+        .into_iter()
+        .flatten()
+    }
+
+    // Only for tests and simulations.
+    pub(crate) fn mock_clone(&self) -> Self {
+        let mut cache = LruCache::new(self.0.cap());
+        for (&origin, entry) in self.0.iter().rev() {
+            cache.put(origin, entry.clone());
+        }
+        Self(cache)
+    }
+}
+
+impl ReceivedCacheEntry {
+    // Limit how big the cache can get if it is spammed
+    // with old messages with random pubkeys.
+    const CAPACITY: usize = 50;
+    // Threshold for the number of duplicates before which a message
+    // is counted as timely towards node's score.
+    const NUM_DUPS_THRESHOLD: usize = 2;
+
+    fn record(&mut self, node: Pubkey, num_dups: usize) {
+        if num_dups == 0 {
+            self.num_upserts = self.num_upserts.saturating_add(1);
+        }
+        // If the message has been timely enough increment node's score.
+        if num_dups < Self::NUM_DUPS_THRESHOLD {
+            let score = self.nodes.entry(node).or_default();
+            *score = score.saturating_add(1);
+        } else if self.nodes.len() < Self::CAPACITY {
+            // Ensure that node is inserted into the cache for later pruning.
+            let _ = self.nodes.entry(node).or_default();
+        }
+    }
+
+    fn prune(
+        self,
+        pubkey: &Pubkey, // This node.
+        origin: &Pubkey, // CRDS value owner.
+        stake_threshold: f64,
+        min_ingress_nodes: usize,
+        stakes: &HashMap<Pubkey, u64>,
+    ) -> impl Iterator<Item = Pubkey> {
+        debug_assert!(self.num_upserts >= ReceivedCache::MIN_NUM_UPSERTS);
+        // Enforce a minimum aggregate ingress stake; see:
+        // https://github.com/solana-labs/solana/issues/3214
+        let min_ingress_stake = {
+            let stake = stakes.get(pubkey).min(stakes.get(origin));
+            (stake.copied().unwrap_or_default() as f64 * stake_threshold) as u64
```

Comment:
any concerns over loss of precision due to `u64` to `f64` conversion?  working purely in `u128`s would be safer

Reply:
I don't think the `u64 -> f64` loss of precision here practically matters.
But I am curious how `u128` is going to help here? 
The issue is conversion to `-> f64`, so I don't see why `u128` would help.

Reply:
I think the precision loss here would be benign.